cudax/include/cuda/experimental/__memory_resource/cuda_async_memory_resource.h

```diff
@@ -0,0 +1,293 @@
+//!    lifetime of the pool exceeds the lifetime of the ``cuda_async_memory_resource``
+//!
+//! @endrst
+class cuda_async_memory_resource
+{
+private:
+  ::cudaMemPool_t __pool_;
+
+  //! @brief Checks whether the passed in alignment is valid
+  //! @param __alignment the alignment to check
+  //! @returns true if \p __alignment is valid
+  _CCCL_NODISCARD static constexpr bool __is_valid_alignment(const size_t __alignment) noexcept
+  {
+    return __alignment <= _CUDA_VMR::default_cuda_malloc_alignment
+        && (_CUDA_VMR::default_cuda_malloc_alignment % __alignment == 0);
+  }
+
+  //! @brief  Returns the default ``cudaMemPool_t`` from the current device
+  //! @throws cuda_error if retrieving the default ``cudaMemPool_t`` fails
+  //! @returns The default memory pool of the current device
+  _CCCL_NODISCARD static ::cudaMemPool_t __get_default_mem_pool(const int __device_id)
+  {
+    _LIBCUDACXX_ASSERT(_CUDA_VMR::__device_supports_pools(__device_id), "cudaMallocAsync not supported");
+
+    ::cudaMemPool_t __pool;
+    _CCCL_TRY_CUDA_API(
+      ::cudaDeviceGetDefaultMemPool, "Failed to call cudaDeviceGetDefaultMemPool", &__pool, __device_id);
+    return __pool;
+  }
+
+public:
+  //! @brief Default constructs the cuda_async_memory_resource using the default \c cudaMemPool_t of the current device
+  //! @throws cuda_error if retrieving the default \c cudaMemPool_t fails
+  cuda_async_memory_resource()
+      : __pool_(__get_default_mem_pool(_CUDA_VMR::__get_current_cuda_device()))
+  {}
+
+  //! @brief Constructs a cuda_async_memory_resource using the default \c cudaMemPool_t of a given deice
+  //! @throws cuda_error if retrieving the default \c cudaMemPool_t fails
+  cuda_async_memory_resource(::cuda::experimental::device_ref __device_ref)
+      : __pool_(__get_default_mem_pool(__device_ref.get()))
+  {}
+
+  cuda_async_memory_resource(int)                   = delete;
+  cuda_async_memory_resource(_CUDA_VSTD::nullptr_t) = delete;
+
+  //! @brief  Constructs the cuda_async_memory_resource from a \c cudaMemPool_t.
+  //! @param __pool The \c cudaMemPool_t used to allocate memory.
+  explicit cuda_async_memory_resource(::cudaMemPool_t __pool) noexcept
+      : __pool_(__pool)
+  {}
+
+  //! @brief  Constructs the cuda_async_memory_resource from a \c cuda_memory_pool by calling native_handle()
+  //! @param __pool The \c cuda_memory_pool used to allocate memory.
+  explicit cuda_async_memory_resource(cuda_memory_pool& __pool) noexcept
+      : __pool_(__pool.native_handle())
+  {}
+
+  //! @brief Allocate device memory of size at least \p __bytes via cudaMallocFromPoolAsync.
+  //! @param __bytes The size in bytes of the allocation.
+  //! @param __alignment The requested alignment of the allocation.
+  //! @throws cuda::std::bad_alloc in case of invalid alignment, or
+  //! @throws cuda::cuda_error If an error code was return by the cuda api call.
```

Comment:
Does the ", or" refer to the next line? I don't think doxygen/breathe will format this correctly.

Reply:
I just checked and this is formatted fine